src/librustc/session/mod.rs

```diff
@@ -1137,9 +1133,20 @@ fn build_session_(
     source_map: Lrc<source_map::SourceMap>,
     driver_lint_caps: FxHashMap<lint::LintId, lint::Level>,
 ) -> Session {
-    let self_profiler =
-        if sopts.debugging_opts.self_profile { Some(Arc::new(PlMutex::new(SelfProfiler::new()))) }
-        else { None };
+    let (self_profiler, profiler_warning) =
+        if sopts.debugging_opts.self_profile {
+            let profiler = SelfProfiler::new();
+            match profiler {
+                Ok(profiler) => {
+                    crate::ty::query::QueryName::register_with_profiler(&profiler);
+                    (Some(Arc::new(profiler)), None)
+                },
+                Err(e) => {
+                    (None, Some(format!("failed to create profiler: {}", e)))
```

Comment:
You should be able to emit warning already here via:
```rust
early_warn(sopts.error_format, &format!("failed to create profiler: {}", e));
```
That should allow to make the control flow simpler.